Failure to resolve exported spread-merged array of cooperating directives between Angular libraries

### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

No

### Description

This reproduction demonstrates an Angular compiler issue where a component imported from another library fails validation if its `imports` array contains another array built with the spread operator.

1. Create workspace

```bash
ng new app --directory=./ --minimal --style=css --ssr=false --zoneless=true --ai-config=none
```

```bash
ng generate library lib-a
```

```bash
ng generate library lib-b
```

```bash
ng generate library lib-c
```

2. Fill `projects/lib-a/src/lib/lib-a.ts`:
```ts
import { Directive } from '@angular/core';

@Directive({ selector: '[baseEntity1]' })
export class BaseEntity1 {}

@Directive({ selector: '[baseEntity2]' })
export class BaseEntity2 {}

export const BaseEntities = [
  BaseEntity1,
  BaseEntity2,
] as const;
```

3. Fill `projects/lib-b/src/lib/lib-b.ts`:
```ts
import { Directive } from '@angular/core';
import { BaseEntities } from 'lib-a';

@Directive({selector: '[moreComplexEntity]'})
export class MoreComplexEntity {}

export const UserFriendlyImport = [...BaseEntities, MoreComplexEntity] as const;
```

4. Fill `projects/lib-с/src/lib/lib-c.ts`:
```ts
import {Component} from '@angular/core';
import {UserFriendlyImport} from 'lib-b';

@Component({
  selector: 'comp-from-library',
  imports: [UserFriendlyImport],
  template: `
    <div baseEntity2>123</div>
    <div moreComplexEntity>123</div>
  `,
})
export class LibraryComponent {}
```

5. Replace all `src/app/app.ts` with the following content:
```ts
import {Component} from '@angular/core';
import {LibraryComponent} from 'lib-c';

@Component({
  selector: 'app-root',
  imports: [LibraryComponent],
  template: `<comp-from-library />`,
})
export class App {}
```

6. Run
```bash
ng build lib-a && ng build lib-b && ng build lib-c && ng build app
```

The application build fails with:
```terminaloutput
✘ [ERROR] NG2012: Component imports must be standalone components, directives, pipes, or must be NgModules. [plugin angular-compiler]

    src/app/app.ts:6:12:
      6 │   imports: [LibraryComponent],
        ╵             ~~~~~~~~~~~~~~~~
```

7. Make these changes inside `projects/lib-b/src/lib/lib-b.ts`
```diff
- export const UserFriendlyImport = [...BaseEntities, MoreComplexEntity] as const;
+ export const UserFriendlyImport = [BaseEntity1, BaseEntity2, MoreComplexEntity] as const;
```

**No error!**
This confirms that the issue is related specifically to using the spread operator in a shared constant that is later consumed via `imports`.

### Please provide a link to a minimal reproduction of the bug

https://github.com/nsbarsukov/ng-import-bug-repro

### Please provide the exception or error you saw

```true
✘ [ERROR] NG2012: Component imports must be standalone components, directives, pipes, or must be NgModules. [plugin angular-compiler]

    src/app/app.ts:6:12:
      6 │   imports: [LibraryComponent],
        ╵             ~~~~~~~~~~~~~~~~
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI       : 21.0.0
Angular           : 21.0.1
Node.js           : 24.10.0
Package Manager   : npm 11.6.1
Operating System  : darwin arm64

┌───────────────────────────┬───────────────────┬───────────────────┐
│ Package                   │ Installed Version │ Requested Version │
├───────────────────────────┼───────────────────┼───────────────────┤
│ @angular/build            │ 21.0.0            │ ^21.0.0           │
│ @angular/cli              │ 21.0.0            │ ^21.0.0           │
│ @angular/common           │ 21.0.1            │ ^21.0.0           │
│ @angular/compiler         │ 21.0.1            │ ^21.0.0           │
│ @angular/compiler-cli     │ 21.0.1            │ ^21.0.0           │
│ @angular/core             │ 21.0.1            │ ^21.0.0           │
│ @angular/forms            │ 21.0.1            │ ^21.0.0           │
│ @angular/platform-browser │ 21.0.1            │ ^21.0.0           │
│ @angular/router           │ 21.0.1            │ ^21.0.0           │
│ ng-packagr                │ 21.0.0            │ ^21.0.0           │
│ rxjs                      │ 7.8.2             │ ~7.8.0            │
│ typescript                │ 5.9.3             │ ~5.9.2            │
│ vitest                    │ 4.0.14            │ ^4.0.8            │
└───────────────────────────┴───────────────────┴───────────────────┘
```

### Anything else?

**Why we use `[...] as const` approach ?**
In [Taiga UI](https://github.com/taiga-family/taiga-ui) (Angular UI library), many apparent "single" components are actually composed from multiple components and directives. With NgModules this complexity was encapsulated: consumers imported one module and were done.

But with standalone components, this changed. Now, library consumers needed to import every individual piece manually — which wasn’t great for DX. IDEs like WebStorm started suggesting dozens of imports, and that quickly got annoying.

To restore a single, ergonomic entry point, we group related declarations into `as const` arrays.

For example:

```ts
export const TuiTextfield = [
  TuiLabel,
  TuiTextfieldComponent,
  TuiTextfieldDirective,
  TuiTextfieldOptionsDirective,
] as const;
```

This keeps the public API compact: the IDE suggests one symbol, while the array still contains all necessary pieces.

More background:  
https://medium.com/angularwave/watching-angular-evolve-taiga-ui-kit-maintainers-perspective-97d2dd56b607#72b2

**Why spread support is important?**<br />

The non-spread workaround:

```diff
- export const UserFriendlyImport = [...BaseEntities, MoreComplexEntity] as const;
+ export const UserFriendlyImport = [BaseEntity1, BaseEntity2, MoreComplexEntity] as const;
```

But in real-world projects with dozens of components and directives, manually listing each one becomes cumbersome and error-prone. The spread operator allows us to maintain modularity and reusability without sacrificing developer experience.

Just explore how we have to violate DRY in Taiga UI now only for textfield components:
* https://github.com/taiga-family/taiga-ui/pull/12671

That PR shows extra duplication just for one type of components. With dozens of similar components, the maintenance overhead becomes significant. Supporting spread-merged arrays in `imports` would allow Angular libraries to stay modular and maintainable without sacrificing DX.

--- a/packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts
+++ b/packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts
@@ -99,10 +99,18 @@ export function validateAndFlattenComponentImports(
 
   for (let i = 0; i < imports.length; i++) {
     const ref = imports[i];
+    let refExpr = expr;
+    if (
+      ts.isArrayLiteralExpression(expr) &&
+      expr.elements.length === imports.length &&
+      !expr.elements.some(ts.isSpreadAssignment)
+    ) {
+      refExpr = expr.elements[i];
+    }
 
     if (Array.isArray(ref)) {
       const {imports: childImports, diagnostics: childDiagnostics} =
-        validateAndFlattenComponentImports(ref, expr, isDeferred);
+        validateAndFlattenComponentImports(ref, refExpr, isDeferred);
       flattened.push(...childImports);
       diagnostics.push(...childDiagnostics);
     } else if (ref instanceof Reference) {
@@ -138,20 +146,16 @@ export function validateAndFlattenComponentImports(
       let diagnosticNode: ts.Node;
       let diagnosticValue: ResolvedValue;
 
-      if (ref instanceof DynamicValue) {
+      // Reporting a diagnostic on the entire array can be noisy, especially if the user has a
+      // large array. Attempt to determine the most accurate position within the `imports` expression to report the
+      // diagnostic on.
+      if (ref instanceof DynamicValue && isWithinExpression(ref.node, expr)) {
+        // Use the dynamic value position itself if it occurs within the `imports` expression.
         diagnosticNode = ref.node;
         diagnosticValue = ref;
-      } else if (
-        ts.isArrayLiteralExpression(expr) &&
-        expr.elements.length === imports.length &&
-        !expr.elements.some(ts.isSpreadAssignment) &&
-        !imports.some(Array.isArray)
-      ) {
-        // Reporting a diagnostic on the entire array can be noisy, especially if the user has a
-        // large array. The most common case is that the array will be static so we can reliably
-        // trace back a `ResolvedValue` back to its node using its index. This allows us to report
-        // the exact node that caused the issue.
-        diagnosticNode = expr.elements[i];
+      } else if (refExpr !== expr) {
+        // The reference comes from a specific element in `expr`, so use that element to report the diagnostic on.
+        diagnosticNode = refExpr;
         diagnosticValue = ref;
       } else {
         diagnosticNode = expr;
@@ -165,6 +169,17 @@ export function validateAndFlattenComponentImports(
   }
 
   return {imports: flattened, diagnostics};
+}
+
+function isWithinExpression(node: ts.Node, expr: ts.Expression): boolean {
+  let current: ts.Node | undefined = node;
+  while (current !== undefined) {
+    if (current === expr) {
+      return true;
+    }
+    current = current.parent;
+  }
+  return false;
 }
 
 /**

--- a/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts
+++ b/packages/compiler-cli/src/ngtsc/partial_evaluator/src/interpreter.ts
@@ -734,13 +734,14 @@ export class StaticInterpreter {
   }
 
   private visitTypeQuery(node: ts.TypeQueryNode, context: Context): ResolvedValue {
-    if (!ts.isIdentifier(node.exprName)) {
+    const exprName = ts.isQualifiedName(node.exprName) ? node.exprName.right : node.exprName;
+    if (!ts.isIdentifier(exprName)) {
       return DynamicValue.fromUnknown(node);
     }
 
-    const decl = this.host.getDeclarationOfIdentifier(node.exprName);
+    const decl = this.host.getDeclarationOfIdentifier(exprName);
     if (decl === null) {
-      return DynamicValue.fromUnknownIdentifier(node.exprName);
+      return DynamicValue.fromUnknownIdentifier(exprName);
     }
 
     const declContext: Context = {...context, ...joinModuleContext(context, node, decl)};

--- a/packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts
+++ b/packages/compiler-cli/src/ngtsc/partial_evaluator/test/evaluator_spec.ts
@@ -425,6 +425,33 @@ runInEachFileSystem(() => {
       expect(local.ownedByModuleGuess).toBeNull();
     });
 
+    // https://github.com/angular/angular/issues/65686
+    it('supports declarations of readonly tuples with class references using qualified names', () => {
+      const tuple = evaluate(
+        `
+        import * as ext from 'external';
+        declare class Local {}
+        declare const x: readonly [typeof ext.External];`,
+        `x`,
+        [
+          {
+            name: _('/node_modules/external/index.d.ts'),
+            contents: 'export declare class External {}',
+          },
+        ],
+      );
+      if (!Array.isArray(tuple)) {
+        return fail('Should have evaluated tuple as an array');
+      }
+      const [external] = tuple;
+      if (!(external instanceof Reference)) {
+        return fail('Should have evaluated `typeof ext.External` to a Reference');
+      }
+      expect(ts.isClassDeclaration(external.node)).toBe(true);
+      expect(external.debugName).toBe('External');
+      expect(external.ownedByModuleGuess).toBe('external');
+    });
+
     it('evaluates tuple elements it cannot understand to DynamicValue', () => {
       const value = evaluate(`declare const x: ['foo', string];`, `x`) as [string, DynamicValue];
 

--- a/packages/compiler-cli/test/ngtsc/ngtsc_spec.ts
+++ b/packages/compiler-cli/test/ngtsc/ngtsc_spec.ts
@@ -2589,11 +2589,69 @@ runInEachFileSystem((os: string) => {
           export class HelloDir {}
 
           const someVar = {} as any;
+          const tuple = [() => {}] as const;
 
           @Component({
             template: '<div hello></div>',
             imports: [
               someVar,
+              HelloDir,
+              'invalid',
+              tuple,
+            ]
+          })
+          export class TestCmp {}
+        `,
+      );
+
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(3);
+      {
+        const message = ts.flattenDiagnosticMessageText(diags[0].messageText, '\n');
+        expect(getDiagnosticSourceCode(diags[0])).toBe('someVar');
+        expect(message).toContain(
+          `'imports' must be an array of components, directives, pipes, or NgModules.`,
+        );
+        expect(message).toContain(`Value is of type '{}'`);
+      }
+      {
+        const message = ts.flattenDiagnosticMessageText(diags[1].messageText, '\n');
+        expect(getDiagnosticSourceCode(diags[1])).toBe(`'invalid'`);
+        expect(message).toContain(
+          `'imports' must be an array of components, directives, pipes, or NgModules.`,
+        );
+        expect(message).toContain(`Value is of type 'string'`);
+      }
+      {
+        const message = ts.flattenDiagnosticMessageText(diags[2].messageText, '\n');
+        expect(getDiagnosticSourceCode(diags[2])).toBe('tuple');
+        expect(message).toContain(
+          `'imports' must be an array of components, directives, pipes, or NgModules.`,
+        );
+        expect(message).toContain(`Value is of type '[(not statically analyzable)]'.`);
+      }
+    });
+
+    it('should report imports diagnostic for declaration file in original expression', () => {
+      env.write(
+        'node_modules/external/index.d.ts',
+        `
+        export declare const UNRESOLVED_ITEM: readonly [unresolved];
+      `,
+      );
+      env.write(
+        'test.ts',
+        `
+          import {Component, Directive} from '@angular/core';
+          import {UNRESOLVED_ITEM as Unresolved} from 'external';
+
+          @Directive({selector: '[hello]'})
+          export class HelloDir {}
+
+          @Component({
+            template: '<div hello></div>',
+            imports: [
+              [Unresolved],
               HelloDir,
             ]
           })
@@ -2606,11 +2664,12 @@ runInEachFileSystem((os: string) => {
         ? ts.flattenDiagnosticMessageText(diags[0].messageText, '\n')
         : '';
       expect(diags.length).toBe(1);
-      expect(getDiagnosticSourceCode(diags[0])).toBe('someVar');
+      expect(diags[0].file!.fileName).toContain('test.ts');
+      expect(getDiagnosticSourceCode(diags[0])).toBe('Unresolved');
       expect(message).toContain(
         `'imports' must be an array of components, directives, pipes, or NgModules.`,
       );
-      expect(message).toContain(`Value is of type '{}'`);
+      expect(message).toContain(`Value is of type '[(not statically analyzable)]'.`);
     });
 
     describe('empty and missing selectors', () => {

